File: tutorials/creating-multi-node-applications/scenario1/multi_ai.py
# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import csv
import os
import sys
from argparse import ArgumentParser

import cupy as cp
import numpy as np
from holoscan.core import Application, Fragment, Operator, OperatorSpec
from holoscan.operators import (
    AJASourceOp,
    FormatConverterOp,
    HolovizOp,
    InferenceOp,
    SegmentationPostprocessorOp,
    VideoStreamReplayerOp,
)
from holoscan.resources import UnboundedAllocator


class DetectionPostprocessorOp(Operator):
    """Example of an operator post processing the tensor from inference component.
    Following the example of tensor_interop.py and ping.py4

    This operator has:
        inputs:  "in"
        outputs: "out"
    """

    def __init__(self, *args, label_dict={}, label_text_size=0.05, scores_threshold=0.3, **kwargs):
        self.label_text_size = label_text_size
        self.scores_threshold = scores_threshold
        self.label_dict = label_dict

        super().__init__(*args, **kwargs)

    def setup(self, spec: OperatorSpec):
        spec.input("in")
        spec.output("out")

    def append_size_to_text_coord(self, text_coord, size):
        # text_coord should be of shape [1, -1, 2]
        # we want to add a third size number to each (x, y)
        # so the text_coord shape [1, -1, 3]
        # the size number determines the text display size in Holoviz
        text_size = np.ones((1, text_coord.shape[1], 1)) * size
        new_text_coord = np.append(text_coord, text_size, 2)
        return new_text_coord.astype(np.float32)

    def compute(self, op_input, op_output, context):
        # Get input message which is a dictionary
        in_message = op_input.receive("in")
        # Convert input to numpy array (using CuPy)
        cp.asarray(in_message.get("inference_output_num_detections")).get()
        output_bboxes = cp.asarray(in_message["inference_output_detection_boxes"]).get()
        output_scores = cp.asarray(in_message["inference_output_detection_scores"]).get()
        output_labels = cp.asarray(in_message["inference_output_detection_classes"]).get()
        # can check the data type of the incoming tensors here
        # print(output_labels.dtype)

        # Threshold output_scores and prune boxes
        ix = output_scores.flatten() >= self.scores_threshold
        has_rect = ix.any()

        output_bboxes = output_bboxes[:, ix, :]  # output_bboxes is of size [1, num_bbox, 4]
        output_labels = output_labels[:, ix].flatten()  # labels is of size [ num_bbox]

        bbox_coords = np.zeros([1, 2, 2], dtype=np.float32)

        if len(self.label_dict) > 0:
            # the label file isn't empty, we want to colorize the bbox and text colors
            bbox_coords = {}
            text_coords = {}
            for label in self.label_dict:
                bbox_coords[label] = np.zeros([1, 2, 2], dtype=np.float32)
                # coords tensor for text to display in Holoviz can be of shape [1, n, 2] or [1, n, 3]
                # with each location having [x,y] coords or [x,y,s] coords where s = size of text
                # to display
                text_coords[label] = np.zeros([1, 1, 2], dtype=np.float32) - 1.0

            if has_rect:
                # there are bboxes and we want to colorize them as well as label text
                for label in self.label_dict:
                    curr_l_ix = output_labels == label

                    if curr_l_ix.any():
                        bbox_coords[label] = np.reshape(output_bboxes[0, curr_l_ix, :], (1, -1, 2))
                        text_coords[label] = self.append_size_to_text_coord(
                            np.reshape(output_bboxes[0, curr_l_ix, :2], (1, -1, 2)),
                            self.label_text_size,
                        )

        else:
            # the label file is empty, just display bboxes in one color
            if has_rect:
                bbox_coords = np.reshape(output_bboxes, (1, -1, 2))

        # Create output message
        out_message = {}
        if len(self.label_dict) > 0:
            # we have split bboxs and text labels into categories
            for label in self.label_dict:
                out_message["rectangles" + str(label)] = bbox_coords[label]
                out_message["label" + str(label)] = text_coords[label]
        else:
            # only transmit the bbox_coords
            out_message["rectangles"] = bbox_coords

        op_output.emit(out_message, "out")


class Fragment1(Fragment):
    def __init__(self, app, name, sample_data_path, source, label_dict):
        super().__init__(app, name)

        self.source = source
        self.label_dict = label_dict
        self.sample_data_path = sample_data_path

    def compose(self):
        # start constructing app
        is_aja = self.source.lower() == "aja"
        SourceClass = AJASourceOp if is_aja else VideoStreamReplayerOp
        source_kwargs = self.kwargs(self.source)
        if self.source == "replayer":
            video_dir = os.path.join(self.sample_data_path, "endoscopy")
            if not os.path.exists(video_dir):
                raise ValueError(f"Could not find video data: {video_dir=}")
            source_kwargs["directory"] = video_dir
        source = SourceClass(self, name=self.source, **source_kwargs)

        in_dtype = "rgba8888" if is_aja else "rgb888"
        pool = UnboundedAllocator(self, name="pool")

        detection_preprocessor = FormatConverterOp(
            self,
            name="detection_preprocessor",
            pool=pool,
            in_dtype=in_dtype,
            **self.kwargs("detection_preprocessor"),
        )
        segmentation_preprocessor = FormatConverterOp(
            self,
            name="segmentation_preprocessor",
            pool=pool,
            in_dtype=in_dtype,
            **self.kwargs("segmentation_preprocessor"),
        )

        model_path_map = {
            "ssd": os.path.join(self.sample_data_path, "ssd_model", "epoch24_nms.onnx"),
            "tool_seg": os.path.join(
                self.sample_data_path,
                "monai_tool_seg_model",
                "model_endoscopic_tool_seg_sanitized_nhwc_in_nchw_out.onnx",
            ),
        }
        for k, v in model_path_map.items():
            if not os.path.exists(v):
                raise RuntimeError(f"Could not find model file: {v}")
        inference_kwargs = self.kwargs("multi_ai_inference")
        inference_kwargs["model_path_map"] = model_path_map

        multi_ai_inference = InferenceOp(
            self,
            name="multi_ai_inference",
            allocator=pool,
            **inference_kwargs,
        )

        detection_postprocessor = DetectionPostprocessorOp(
            self,
            name="detection_postprocessor",
            label_dict=self.label_dict,
            allocator=UnboundedAllocator(self, name="allocator"),
            **self.kwargs("detection_postprocessor"),
        )

        segmentation_postprocessor = SegmentationPostprocessorOp(
            self,
            name="segmentation_postprocessor",
            allocator=pool,
            **self.kwargs("segmentation_postprocessor"),
        )

        # connect the input each pre-processor
        if is_aja:
            self.add_flow(source, detection_preprocessor, {("video_buffer_output", "")})
            self.add_flow(source, segmentation_preprocessor, {("video_buffer_output", "")})

        else:
            self.add_flow(source, detection_preprocessor)
            self.add_flow(source, segmentation_preprocessor)

        # connect all pre-processor outputs to the inference operator
        for op in [detection_preprocessor, segmentation_preprocessor]:
            self.add_flow(op, multi_ai_inference, {("", "receivers")})

        # connect the inference output to the postprocessors
        self.add_flow(multi_ai_inference, detection_postprocessor, {("transmitter", "in")})
        self.add_flow(multi_ai_inference, segmentation_postprocessor, {("transmitter", "")})


class Fragment2(Fragment):
    def __init__(self, app, name, label_dict):
        super().__init__(app, name)

        self.label_dict = label_dict

    def compose(self):
        # Holoviz
        holoviz_tensors = [dict(name="", type="color"), dict(name="out_tensor", type="color_lut")]
        if len(self.label_dict) > 0:
            for label in self.label_dict:
                color = self.label_dict[label]["color"]
                color.append(1.0)
                text = [self.label_dict[label]["text"]]
                holoviz_tensors.append(
                    dict(
                        name="rectangles" + str(label),
                        type="rectangles",
                        opacity=0.7,
                        line_width=4,
                        color=color,
                    )
                )
                holoviz_tensors.append(
                    dict(
                        name="label" + str(label), type="text", opacity=0.7, color=color, text=text
                    )
                )
        else:
            holoviz_tensors.append(
                dict(
                    name="rectangles",
                    type="rectangles",
                    opacity=0.7,
                    line_width=4,
                    color=[1.0, 0.0, 0.0, 1.0],
                )
            )
        pool = UnboundedAllocator(self, name="pool")
        holoviz = HolovizOp(
            self, allocator=pool, name="holoviz", tensors=holoviz_tensors, **self.kwargs("holoviz")
        )

        # add operators

        self.add_operator(holoviz)


class MultiAIDetectionSegmentation(Application):
    def __init__(self, data, source="replayer", labelfile=""):
        super().__init__()

        if data == "none":
            data = os.environ.get("HOLOHUB_DATA_PATH", "../data")

        # set name
        self.name = "Multi AI App"

        # Optional parameters affecting the graph created by compose.
        source = source.lower()
        if source not in ["replayer", "aja"]:
            raise ValueError(f"unsupported source: {source}. Please use 'replayer' or 'aja'.")
        self.source = source
        self.label_dict = self.get_label_dict(labelfile)
        self.sample_data_path = data

    def compose(self):
        fragment1 = Fragment1(
            self,
            name="fragment1",
            source=self.source,
            sample_data_path=self.sample_data_path,
            label_dict=self.label_dict,
        )
        fragment2 = Fragment2(self, name="fragment2", label_dict=self.label_dict)

        # Connect the two fragments
        # We can skip the "out" and "in" suffixes, as they are the default
        source_output = (
            self.source + ".video_buffer_output"
            if self.source.lower() == "aja"
            else self.source + ".output"
        )
        self.add_flow(fragment1, fragment2, {(source_output, "holoviz.receivers")})
        self.add_flow(fragment1, fragment2, {("detection_postprocessor.out", "holoviz.receivers")})
        self.add_flow(
            fragment1, fragment2, {("segmentation_postprocessor.out_tensor", "holoviz.receivers")}
        )

    def get_label_dict(self, labelfile):
        # construct the labels dictionary if the commandline arg for labelfile isn't empty
        label_dict = {}
        if labelfile != "":
            assert os.path.isfile(labelfile)
            with open(labelfile, newline="") as labelcsv:
                csvreader = csv.reader(labelcsv)
                for row in csvreader:
                    # assume each row looks like: 1, "Grasper", 1.0, 0.0, 1.0
                    label_dict[int(row[0])] = {}
                    label_dict[int(row[0])]["text"] = str(row[1])
                    label_dict[int(row[0])]["color"] = [float(row[2]), float(row[3]), float(row[4])]
        return label_dict


if __name__ == "__main__":

    parser = ArgumentParser(description="Multi-AI Detection Segmentation application.")
    parser.add_argument(
        "-s",
        "--source",
        choices=["replayer", "aja"],
        default="replayer",
        help=(
            "If 'replayer', replay a prerecorded video. If 'aja' use an AJA "
            "capture card as the source (default: %(default)s)."
        ),
    )
    parser.add_argument(
        "-c",
        "--config",
        default="none",
        help=("Set config path to override the default config file location"),
    )
    parser.add_argument(
        "-d",
        "--data",
        default="none",
        help=("Set the data path"),
    )
    parser.add_argument(
        "-l",
        "--labelfile",
        default="none",
        help=(
            "Optional arg for a csv file path containing the class labels. There "
            "should be 5 columns: [label value, label text, red, green, blue] "
            "where label value = the model output value for each class to display and should be int, "
            "label text = the text to display for this label, "
            "red/green/blue should = values between 0 and 1 for the display color "
            "for this class."
        ),
    )
    apps_argv = Application().argv
    args = parser.parse_args(apps_argv[1:])
    if args.config == "none":
        config_file = os.path.join(
            os.path.dirname(__file__),
            "../../../applications/multiai_endoscopy/python/multi_ai.yaml",
        )
    else:
        config_file = args.config

    if args.labelfile == "none":
        labelfile = os.path.join(
            os.path.dirname(__file__),
            "../../../applications/multiai_endoscopy/python/endo_ref_data_labels.csv",
        )
    else:
        labelfile = args.labelfile

    app = MultiAIDetectionSegmentation(data=args.data, source=args.source, labelfile=labelfile)
    app.config(config_file)
    print("sys.argv:", sys.argv)
    print("app.argv:", app.argv)
    app.run()

File: applications/multiai_endoscopy/cpp/post-proc-matx-cpu/multi_ai.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2022-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <getopt.h>

#include <cuda_runtime.h>
#include <holoscan/holoscan.hpp>
#include <holoscan/operators/video_stream_replayer/video_stream_replayer.hpp>
#include <holoscan/operators/format_converter/format_converter.hpp>
#include <holoscan/operators/inference/inference.hpp>
#include <holoscan/operators/segmentation_postprocessor/segmentation_postprocessor.hpp>
#include <holoscan/operators/holoviz/holoviz.hpp>

#ifdef AJA_SOURCE
#include <aja_source.hpp>
#endif

#include "gxf/std/tensor.hpp"
#include "matx.h"

#define CUDA_TRY(stmt)                                                                  \
  {                                                                                     \
    cudaError_t cuda_status = stmt;                                                     \
    if (cudaSuccess != cuda_status) {                                                   \
      HOLOSCAN_LOG_ERROR("Runtime call {} in line {} of file {} failed with '{}' ({})", \
                         #stmt,                                                         \
                         __LINE__,                                                      \
                         __FILE__,                                                      \
                         cudaGetErrorString(cuda_status),                               \
                         static_cast<int>(cuda_status));                                \
      throw std::runtime_error("Unable to copy device to host");                        \
    }                                                                                   \
  }

// Debug prints tensor content.
void print_tensor(const std::shared_ptr<holoscan::Tensor>& tensor, size_t n_printouts = 20) {
  HOLOSCAN_LOG_INFO("==========TENSOR INFO=============");

  auto device = tensor->device();
  auto dtype = tensor->dtype();
  auto shape = tensor->shape();
  auto strides = tensor->strides();
  size_t size = tensor->size();
  auto ndim = tensor->ndim();
  auto itemsize = tensor->itemsize();

  HOLOSCAN_LOG_INFO("device.device_type={}, device.device_id={}",
                    static_cast<int>(device.device_type),
                    device.device_id);
  HOLOSCAN_LOG_INFO(
      "dtype.code={}, dtype.bits={}, dtype.lanes={}", dtype.code, dtype.bits, dtype.lanes);
  for (int i = 0; i < shape.size(); ++i) {
    HOLOSCAN_LOG_INFO("shape[{}]={}", i, shape[i]);
  }
  for (int i = 0; i < strides.size(); ++i) {
    HOLOSCAN_LOG_INFO("strides[{}]={}", i, strides[i]);
  }
  HOLOSCAN_LOG_INFO("size={}", size);
  HOLOSCAN_LOG_INFO("ndim={}", ndim);
  HOLOSCAN_LOG_INFO("itemsize={}", itemsize);

  size_t nbytes = tensor->nbytes();
  HOLOSCAN_LOG_INFO("nbytes={}", nbytes);
  std::vector<float> data(nbytes);
  CUDA_TRY(cudaMemcpy(data.data(), tensor->data(), nbytes, cudaMemcpyDeviceToHost));

  for (size_t i = 0; i < std::min(size, n_printouts); i++) {
    HOLOSCAN_LOG_INFO("data[{}]={:f}", i, data[i]);
  }

  CUDA_TRY(cudaMemcpy(tensor->data(), data.data(), nbytes, cudaMemcpyHostToDevice));

  HOLOSCAN_LOG_INFO("==================================");
}

// Copies device Holoscan tensor data to host vector
template <typename T>
std::vector<T> copy_device2vec(const std::shared_ptr<holoscan::Tensor>& in) {
  size_t nbytes = in->nbytes();
  std::vector<T> out = std::vector<T>(nbytes/sizeof(T));

  CUDA_TRY(cudaMemcpy(out.data(), in->data(), nbytes, cudaMemcpyDeviceToHost));

  return out;
}

namespace holoscan::ops {

// Operator for post-processesing inference output
class DetectionPostprocessorOp : public Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(DetectionPostprocessorOp)

  DetectionPostprocessorOp() = default;

  void setup(OperatorSpec& spec) override {
    spec.input<gxf::Entity>("in");
    spec.output<gxf::Entity>("out");
    spec.param(
      scores_threshold_,
      "scores_threshold",
      "Scores Threshold",
      "Threshold NMS scores by this value",
      0.3f);
    spec.param(
      label_names_,
      "label_names",
      "Label Names",
      "List of label names",
      std::vector<std::string>());
  }

  void compute(InputContext& op_input, OutputContext& op_output,
               ExecutionContext& context) override {
    // Get input message and make output message
    auto in_message = op_input.receive<gxf::Entity>("in").value();
    auto out_message = gxf::Entity::New(&context);

    // Get total number of detections
    auto Nd = in_message.get<Tensor>("inference_output_num_detections");
    // Copy to host
    auto Ndh = copy_device2vec<int>(Nd);

    if (Ndh[0] == 0) {
      // Check if zero detections, in which case we return empty messages

      for (int k = 0; k < label_names_.get().size(); k++) {
        // Create empty messages for each label
        auto label = label_names_.get()[k];

        auto boxes_mx = matx::make_tensor<float>({1, 2, 2});
        auto coords_mx = matx::make_tensor<float>({1, 1, 2});
        (coords_mx = -1.0).run(matx::SingleThreadHostExecutor());

        // MatX to Holoscan tensor
        auto boxes_hs = std::make_shared<holoscan::Tensor>(boxes_mx.GetDLPackTensor());
        auto coords_hs = std::make_shared<holoscan::Tensor>(coords_mx.GetDLPackTensor());

        out_message.add(boxes_hs, (label + "_rectangle").c_str());
        out_message.add(coords_hs, (label + "_label").c_str());
      }

      op_output.emit(out_message, "out");

      return;
    }

    // Boxes detected, get boxes, scores and labels
    auto boxesh = in_message.get<Tensor>("inference_output_detection_boxes");  // (1, num_boxes, 4)
    auto scoresh = in_message.get<Tensor>("inference_output_detection_scores");  // (1, num_boxes)
    auto labelsh = in_message.get<Tensor>("inference_output_detection_classes");  // (1, num_boxes)
    int32_t Nb = scoresh->shape()[1];  // Number of boxes

    // Copy to host
    auto boxes = copy_device2vec<float>(boxesh);
    auto scores = copy_device2vec<float>(scoresh);
    auto labels = copy_device2vec<int>(labelsh);

    // Holoscan tensors to MatX tensors
    auto scores_mx = matx::make_tensor<float>(scores.data(), {1, Nb});
    auto boxes_mx = matx::make_tensor<float>(boxes.data(), {1, Nb, 4});
    auto labels_mx = matx::make_tensor<int>(labels.data(), {1, Nb});

    // Find box indices with a score larger than threshold
    auto ix_mx = matx::make_tensor<int>({Nb});
    auto Ns = matx::make_tensor<int>({});  // Number of boxes above threshold
    (matx::mtie(ix_mx, Ns) = matx::find_idx(scores_mx, matx::GTE{scores_threshold_.get()})).run(
                                         matx::SingleThreadHostExecutor());

    // Get boxes and labels corresponding to indices
    auto boxes_ix_mx = matx::make_tensor<float>({1, Ns(), 4});
    auto labels_ix_mx = matx::make_tensor<int>({1, Ns()});
    auto ixs_mx = ix_mx.Slice({0}, {Ns()});
    (boxes_ix_mx = matx::remap<1>(boxes_mx, ixs_mx)).run(matx::SingleThreadHostExecutor());
    (labels_ix_mx = matx::remap<1>(labels_mx, ixs_mx)).run(matx::SingleThreadHostExecutor());

    // Get x0,y0 coordinates from boxes
    auto coords_mx = matx::make_tensor<float>({1, Ns(), 2});
    (coords_mx = boxes_ix_mx.Slice({0, 0, 0}, {matx::matxEnd, matx::matxEnd, 2})).run(
                                                                matx::SingleThreadHostExecutor());

    // Create messages for each label
    auto Nl = matx::make_tensor<int>({});  // Number of label boxes
    for (int k = 0; k < label_names_.get().size(); k++) {
      // Loop over label name and index
      auto label = label_names_.get()[k];

      // Find boxes equal to label index
      (matx::mtie(ixs_mx, Nl) =
          matx::find_idx(labels_ix_mx, matx::EQ{k + 1})).run(matx::SingleThreadHostExecutor());

      if (Nl() > 0) {
        // Label has boxes, create output messages
        auto ixl_mx = ixs_mx.Slice({0}, {Nl()});

        // Get boxes and labels corresponding to indices
        auto boxesl_mx = matx::make_tensor<float>({1, Nl(), 4});
        auto coordsl_mx = matx::make_tensor<float>({1, Nl(), 2});
        (boxesl_mx = matx::remap<1>(boxes_ix_mx, ixl_mx)).run(matx::SingleThreadHostExecutor());
        (coordsl_mx = matx::remap<1>(coords_mx, ixl_mx)).run(matx::SingleThreadHostExecutor());

        // Reshape boxes to Holoviz shape
        auto boxesls_mx = boxesl_mx.View({1, 2 * Nl(), 2});

        // MatX to Holoscan tensor
        auto boxes_hs = std::make_shared<holoscan::Tensor>(boxesls_mx.GetDLPackTensor());
        auto coords_hs = std::make_shared<holoscan::Tensor>(coordsl_mx.GetDLPackTensor());

        out_message.add(boxes_hs, (label + "_rectangle").c_str());
        out_message.add(coords_hs, (label + "_label").c_str());

      } else {
        // Label has no boxes, create empty output messages
        auto boxesl_mx = matx::make_tensor<float>({1, 2, 2});
        auto coordsl_mx = matx::make_tensor<float>({1, 1, 2});
        (coordsl_mx = -1.0).run(matx::SingleThreadHostExecutor());

        // MatX to Holoscan tensor
        auto boxes_hs = std::make_shared<holoscan::Tensor>(boxesl_mx.GetDLPackTensor());
        auto coords_hs = std::make_shared<holoscan::Tensor>(coordsl_mx.GetDLPackTensor());

        out_message.add(boxes_hs, (label + "_rectangle").c_str());
        out_message.add(coords_hs, (label + "_label").c_str());
      }
    }

    // Emit output message
    op_output.emit(out_message, "out");
  };

 private:
  Parameter<float> scores_threshold_;
  Parameter<std::vector<std::string>> label_names_;
};

}  // namespace holoscan::ops

class App : public holoscan::Application {
 public:
  void set_source(const std::string& source) {
    if (source == "aja") {
      is_aja_source_ = true;
    }
  }

  void set_datapath(const std::string& path) {
     datapath = path;
  }

  void compose() override {
    using namespace holoscan;

    /* Allocator */
    std::shared_ptr<Resource> pool = make_resource<UnboundedAllocator>("pool");

    /* Source */
    std::shared_ptr<Operator> source;
    if (is_aja_source_) {
#ifdef AJA_SOURCE
      source = make_operator<ops::AJASourceOp>("aja", from_config("aja"));
#else
      throw std::runtime_error(
          "AJA is requested but not available. Please enable AJA at build time.");
#endif
    } else {
      source = make_operator<ops::VideoStreamReplayerOp>("replayer", from_config("replayer"),
                                                  Arg("directory", datapath + "/endoscopy"));
    }

    /* preprocessor segmentations */
    auto in_dtype = is_aja_source_ ? std::string("rgba8888") : std::string("rgb888");
    auto segmentation_preprocessor = make_operator<ops::FormatConverterOp>(
                                                          "segmentation_preprocessor",
                                                          from_config("segmentation_preprocessor"),
                                                          Arg("in_dtype") = in_dtype,
                                                          Arg("pool") = pool);

    /* preprocessor boxes */
    auto detection_preprocessor = make_operator<ops::FormatConverterOp>(
        "detection_preprocessor",
        from_config("detection_preprocessor"),
        Arg("in_tensor_name", std::string(is_aja_source_ ? "source_video" : "")),
        Arg("in_dtype") = in_dtype,
        Arg("pool") = pool);

    /* inference */
    ops::InferenceOp::DataMap model_path_map;
    model_path_map.insert("ssd", datapath + "/ssd_model/epoch24_nms.onnx");
    model_path_map.insert("tool_seg", datapath +
     "/monai_tool_seg_model/model_endoscopic_tool_seg_sanitized_nhwc_in_nchw_out.onnx");
    auto inference = make_operator<ops::InferenceOp>(
        "inference", from_config("inference"),
        Arg("model_path_map", model_path_map),
        Arg("allocator") = pool);

    /* segmentation postprocessor */
    auto segmentation_postprocessor = make_operator<ops::SegmentationPostprocessorOp>(
        "segmentation_postprocessor", from_config("segmentation_postprocessor"),
                                                        Arg("allocator") = pool);

    /* detection postprocessor */
    auto detection_postprocessor = make_operator<ops::DetectionPostprocessorOp>(
        "detection_postprocessor", from_config("detection_postprocessor"),
                                               Arg("allocator") = pool);

    /* visualizer */
    auto holoviz = make_operator<ops::HolovizOp>(
        "holoviz", from_config("holoviz"), Arg("allocator") = pool);

    /* Define flow */

    // Source
    if (is_aja_source_) {
      add_flow(source, holoviz, {{"video_buffer_output", "receivers"}});
      add_flow(source, detection_preprocessor, {{"video_buffer_output", ""}});
      add_flow(source, segmentation_preprocessor, {{"video_buffer_output", ""}});
    } else {
      add_flow(source, holoviz, {{"", "receivers"}});
      add_flow(source, detection_preprocessor);
      add_flow(source, segmentation_preprocessor);
    }

    // Detection
    add_flow(detection_preprocessor, inference, {{"", "receivers"}});
    add_flow(inference, detection_postprocessor, {{"transmitter", "in"}});
    add_flow(detection_postprocessor, holoviz, {{"out", "receivers"}});

    // Segmentation
    add_flow(segmentation_preprocessor, inference, {{"", "receivers"}});
    add_flow(inference, segmentation_postprocessor, {{"transmitter", ""}});
    add_flow(segmentation_postprocessor, holoviz, {{"", "receivers"}});
  }

 private:
  bool is_aja_source_ = false;
  std::string datapath = "data";
};

/** Helper function to parse the command line arguments */
bool parse_arguments(int argc, char** argv, std::string& config_name, std::string& data_path) {
  static struct option long_options[] = {
      {"data",    required_argument, 0,  'd' },
      {0,         0,                 0,  0 }
  };

  while (int c = getopt_long(argc, argv, "d",
                   long_options, NULL))  {
    if (c == -1 || c == '?') break;

    switch (c) {
      case 'd':
        data_path = optarg;
        break;
      default:
        std::cout << "Unknown arguments returned: " << c << std::endl;
        return false;
    }
  }

  if (optind < argc) {
    config_name = argv[optind++];
  }
  return true;
}

int main(int argc, char** argv) {
  auto app = holoscan::make_application<App>();

  // Parse the arguments
  std::string data_path = "";
  std::string config_name = "";
  if (!parse_arguments(argc, argv, config_name, data_path)) {
    return 1;
  }

  if (config_name != "") {
    app->config(config_name);
  } else {
    auto config_path = std::filesystem::canonical(argv[0]).parent_path();
    config_path += "/app_config.yaml";
    app->config(config_path);
  }

  auto source = app->from_config("source").as<std::string>();
  app->set_source(source);
  if (data_path != "") app->set_datapath(data_path);
  app->run();

  return 0;
}
